core/src/main/java/com/zextras/carbonio/files/message_broker/consumers/KvChangedConsumer.java

```diff
@@ -0,0 +1,98 @@
+import com.zextras.carbonio.files.dal.repositories.interfaces.FileVersionRepository;
+import com.zextras.carbonio.files.dal.repositories.interfaces.NodeRepository;
+import com.zextras.carbonio.message_broker.config.EventConfig;
+import com.zextras.carbonio.message_broker.consumer.BaseConsumer;
+import com.zextras.carbonio.message_broker.events.generic.BaseEvent;
+import com.zextras.carbonio.message_broker.events.services.service_discover.KvChanged;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class KvChangedConsumer extends BaseConsumer {
+
+  private static final Logger logger = LoggerFactory.getLogger(KvChangedConsumer.class);
+
+  private final NodeRepository nodeRepository;
+  private final FileVersionRepository fileVersionRepository;
+
+  public KvChangedConsumer(NodeRepository nodeRepository, FileVersionRepository fileVersionRepository) {
+    this.nodeRepository = nodeRepository;
+    this.fileVersionRepository = fileVersionRepository;
+  }
+
+  @Override
+  protected EventConfig getEventConfig() {
+    return EventConfig.KV_CHANGED;
+  }
+
+  @Override
+  public void doHandle(BaseEvent baseMessageBrokerEvent) {
+    KvChanged kvChanged = (KvChanged) baseMessageBrokerEvent;
+    logger.info("Received KvChanged({}, {})", kvChanged.getKey(), kvChanged.getValue());
+
+    // switch for future use cases with KVs
+    switch (kvChanged.getKey()) {
+      case "carbonio-files/max-number-of-versions":
+        handleMaxVersionNumerChanged(Integer.parseInt(kvChanged.getValue()));
+        break;
+      default:
+        logger.info("Event was ignored");
+        break;
+    }
+  }
+
+  /**
+   * Check all nodes' versions and deletes the least recent ones to match the new max version number.
+   * There are a few exceptions: if a file version is marked as keep forever it does not delete it (this of course
+   * can cause a situation where there are more versions that new max version number); it also does not delete the current
+   * (last) version of a node ever, possibly resulting in the same scenario of having more versions than maximum.
+   */
+  private void handleMaxVersionNumerChanged(int newMaxVersionNumber) {
+    logger.info("Handling new max version number");
+    List<Node> allFiles = nodeRepository.findAllNodesFiles();
+
+    /*
+     I don't know how to make this more efficient (at least just a query once retrieved all files) without
+     writing an unreadable mess of a query with subqueries that essentially do the same thing;
+     since it's a rare operation I think it can stay like this for readability
+    */
+    for (Node node : allFiles) {
+      logger.info("Checking node {}", node.getFullName());
+      List<FileVersion> filesToCheck = fileVersionRepository.getFileVersions(node.getId(), List.of(FileVersionSort.VERSION_ASC));
+
+      if (filesToCheck.size() > newMaxVersionNumber) {
+        trimFileVersions(filesToCheck, newMaxVersionNumber);
+      }
+    }
```

Comment:
I don't like this solution because, as you said, it is very inefficient. I propose to create another method in the `FileVersionRepository` returning all the fileversion to "trim". The for loop remains only to call the `trimFileVersions` on each node.


```java
  /**
   * Retrieves all @link FileVersion}s belonging to a {@link Node} representing a file having a
   * number of version greater than a specific value.
   *
   * @param maxNumberOfVersions is a <code>int</code> representing the max number of version a
   *                            {@link Node} should have
   * @return a {@link Map} where the key is the node identifier and the value contains a
   * {@link List} of {@link FileVersion}s linked to that node ordered by version descending.
   */
  public Map<String, List<FileVersion>> getFileVersionsRelatedToNodesHavingVersionsGreaterThan(
    int maxNumberOfVersions) {
    return mDB.getEbeanDatabase()
      .find(FileVersion.class)
      .fetch("node")
      .having()
      .gt("node.mCurrentVersion", maxNumberOfVersions)
      .orderBy()
      .desc(Db.FileVersion.VERSION)
      .setMapKey(Db.FileVersion.NODE_ID)
      .findList()
      .stream()
      .collect(Collectors.groupingBy(FileVersion::getNodeId));
  }
```